### DISCUSSION TITLE
Register additional tab to plugin model

### DISCUSSION THREAD

**Discussion by Kani999** [netbox-community/netbox#11641] [Q&A] [answered]
Hello,

i'm trying to register additional tab ( https://docs.netbox.dev/en/stable/plugins/development/views/#additional-tabs ) to my plugin model.

Examples:
- Plugin model `Contract`
```python
@register_model_view(Contract, 'consoleports', path='console-ports')
class ContractConsolePortsView(DeviceComponentsView):
    child_model = ConsolePort
    table = DeviceConsolePortTable
    filterset = ConsolePortFilterSet
    template_name = 'dcim/device/consoleports.html',
    tab = ViewTab(
        label=('Console Ports'),
        badge=0,
        weight=550,
        hide_if_empty=True
    )
```

But the tab does not render on the object page. 
The `registry` seems fine 

```python
ipdb> from pprint import pprint
ipdb> pprint(registry['views']['inventory_monitor']['contract'])
[{'kwargs': {'model': <class 'inventory_monitor.models.Contract'>},
  'name': 'journal',
  'path': 'journal',
  'view': 'netbox.views.generic.ObjectJournalView'},
 {'kwargs': {'model': <class 'inventory_monitor.models.Contract'>},
  'name': 'changelog',
  'path': 'changelog',
  'view': 'netbox.views.generic.ObjectChangeLogView'},
 {'kwargs': {},
  'name': 'contract-attachment_list',
  'path': 'contract-attachment_list',
  'view': <class 'netbox_attachments.template_content.create_tab_view.<locals>.View'>},
 {'kwargs': {},
  'name': '',
  'path': '',
  'view': <class 'inventory_monitor.views.ContractView'>},
 {'kwargs': {},
  'name': 'consoleports',
  'path': 'console-ports',
  'view': <class 'inventory_monitor.views.ContractConsolePortsView'>}]
  ```
The only tab which is displayed is `Changelog`

![image](https://user-images.githubusercontent.com/17337078/216300745-9790df7e-508e-4493-bdfa-c49163bd4554.png)

Registering tabs to "Core" models, like "Device" works fine.

![image](https://user-images.githubusercontent.com/17337078/216300976-b699dc4e-16e0-4750-9973-47a3394cd12d.png)

Are there some caveats, that I cannot add tabs to my plugin? 
Or should I add something to my models, views, etc? 

---
I create a discussion on slack, where I tried several things.
https://netdev-community.slack.com/archives/C03Q7A8C0TS/p1675324918322649

**Comment by AcidSailor** [comment, marked as answer]
I didn't check the slack discussion, but eventually came up with the solution, so posting it here:

1. You should use this in `urls.py` for your custom plugin (basically, `get_model_urls` seems to load all urls from `registry`)
```python3
include(get_model_urls(<app_label>, <model_name>)))
```
For example, how it's done in core models:
https://github.com/netbox-community/netbox/blob/0e34fba92223348e0bf4375b8d380324ff5e1beb/netbox/dcim/urls.py#L191
2. Or you also can do it "manually", for me it was:
`views.py`
```python3
@register_model_view(
    models.ProviderService,
    name="contacts",
)
class ProviderServiceContactsView(ObjectContactsView):
    queryset = models.ProviderService.objects.all()

```

`urls.py`
```python3
urlpatterns = (
    path("provider-services/<int:pk>/",
         views.ProviderServiceView.as_view(), name="providerservice"),
    path("provider-services/<int:pk>/edit/", views.ProviderServiceEditView.as_view(),
         name="providerservice_edit"),
    path("provider-services/<int:pk>/delete/", views.ProviderServiceDeleteView.as_view(),
         name="providerservice_delete"),
    path("provider-services/<int:pk>/changelog/", ObjectChangeLogView.as_view(),
         kwargs={"model": models.ProviderService}, name="providerservice_changelog"),

    # Without this line, custom tab was not working:
    path("provider-services/<int:pk>/contacts/", views.ProviderServiceContactsView.as_view(),
         name="providerservice_contacts"),
)
```

**Reply by Kani999** [reply]
Using `get_model_urls` works well for me.  

It automatically adds Changelog, Journal, and other views registered with `@register_model_view`.  

Additionally, if the `netbox-attachments` plugin is installed and configured to display as an `addition_tab`, it will also appear.  

Example:  

```python
from django.urls import include, path
from utilities.urls import get_model_urls

...

path(
    "segments/<int:pk>/",
    include(get_model_urls("cesnet_service_path_plugin", "segment")),
),

...
```
